### PR TITLE
api: Add TitlePart

### PR DESCRIPTION
**Pull request by kezz**
This PR is an alternative to #320. With this PR, you would send individual title parts like:
```java
audience.showTitlePart(TitlePart.TITLE, text("hello!"));
audience.showTitlePart(TitlePart.TIMES, times(0, 100, 15));
```

### CODE REVIEW

File: api/src/main/java/net/kyori/adventure/audience/ForwardingAudience.java [hunk 1]
```diff
@@ -112,6 +113,11 @@ default void showTitle(final @NotNull Title title) {
     for (final Audience audience : this.audiences()) audience.showTitle(title);
   }
 
+  @Override
+  default <T> void showTitlePart(final @NotNull TitlePart<T> titlePart, @NotNull final T content) {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/audience/ForwardingAudience.java:117]: Perhaps `sendTitlePart`?

Comment (kezz) [reply api/src/main/java/net/kyori/adventure/audience/ForwardingAudience.java]: Yeah I think this does make more sense, more in keeping with the rest of the API

**Review by Citymonstret — APPROVED** [review]

**Comment by kezz** [comment]
Just spotted some incorrect `@since` tags so have updated those and rebased to latest. For reference, the client doesn't care about the ordering of the packets, so the replacement implementation for the `showTitle` method is also fine.

**Review by Proximyst — APPROVED** [review]

File: api/src/main/java/net/kyori/adventure/audience/Audience.java [hunk 2]
```diff
@@ -403,7 +404,24 @@ default void sendPlayerListHeaderAndFooter(final @NotNull Component header, fina
...
+
+  /**
+   * Shows a part of a title.
+   *
+   * @param titlePart the part
+   * @param content the content
+   * @param <T> the type of the content of the part
+   * @since 4.9.0
+   */
+  default <T> void sendTitlePart(final @NotNull TitlePart<T> titlePart, final @NotNull T content) {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/audience/Audience.java:424]:
```suggestion
  default <T> void sendTitlePart(final @NotNull TitlePart<T> part, final @NotNull T value) {
```

File: api/src/main/java/net/kyori/adventure/audience/ForwardingAudience.java [hunk 3]
```diff
@@ -109,8 +109,8 @@ default void sendPlayerListHeaderAndFooter(final @NotNull Component header, fina
   }
 
   @Override
-  default void showTitle(final @NotNull Title title) {
-    for (final Audience audience : this.audiences()) audience.showTitle(title);
+  default <T> void sendTitlePart(final @NotNull TitlePart<T> titlePart, @NotNull final T content) {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/audience/ForwardingAudience.java:112]: and here

File: api/src/main/java/net/kyori/adventure/audience/ForwardingAudience.java [hunk 4]
```diff
@@ -237,8 +237,8 @@ default void sendPlayerListHeaderAndFooter(final @NotNull Component header, fina
     }
 
     @Override
-    default void showTitle(final @NotNull Title title) {
-      this.audience().showTitle(title);
+    default <T> void sendTitlePart(final @NotNull TitlePart<T> titlePart, @NotNull final T content) {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/audience/ForwardingAudience.java:240]: and here

File: api/src/main/java/net/kyori/adventure/title/TitlePart.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+public interface TitlePart<T> {
+  /**
+   * A title part of a title.
+   *
+   * @since 4.9.0
+   */
+  @NotNull TitlePart<Component> TITLE = new TitlePart<Component>() {
+    @Override
+    public String toString() {
+      return "TitleTitlePart";
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/title/TitlePart.java:47]:
```suggestion
      return "TitlePart.TITLE";
```

File: api/src/main/java/net/kyori/adventure/title/TitlePart.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+  /**
+   * The subtitle part of a title.
+   *
+   * @since 4.9.0
+   */
+  @NotNull TitlePart<Component> SUBTITLE = new TitlePart<Component>() {
+    @Override
+    public String toString() {
+      return "SubtitleTitlePart";
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/title/TitlePart.java:59]:
```suggestion
      return "TitlePart.SUBTITLE";
```

File: api/src/main/java/net/kyori/adventure/title/TitlePart.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+  /**
+   * The times part of a title.
+   *
+   * @since 4.9.0
+   */
+  @NotNull TitlePart<Title.Times> TIMES = new TitlePart<Title.Times>() {
+    @Override
+    public String toString() {
+      return "TimesTitlePart";
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/title/TitlePart.java:71]:
```suggestion
      return "TitlePart.TIMES";
```

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by zml2008** [comment]
Could you add some API to this to map from a `Title` to and from its component `TitlePart`s?

something like `Title.part(TitlePart)` and `Title.title(PartValuePair...)` or something along those lines?

**Comment by kezz** [comment]
Does anyone have any thoughts about `sendTitlePart` vs `showTitlePart`?

**Comment by kashike** [comment]
Probably `sendTitlePart`

**Review by zml2008 — APPROVED** [review]

File: api/src/main/java/net/kyori/adventure/title/TitleImpl.java [hunk 8]
```diff
@@ -58,6 +59,20 @@
...
+  public <T> @UnknownNullability T part(final @NotNull TitlePart<T> part) {
+    if (part == TitlePart.TITLE) {
+      return (T) this.title;
+    } else if (part == TitlePart.SUBTITLE) {
+      return (T) this.subtitle;
+    } else if (part == TitlePart.TIMES) {
+      return (T) this.times;
+    }
+
+    throw new IllegalArgumentException("Unknown part.");
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/title/TitleImpl.java:73]:
```suggestion
    throw new IllegalArgumentException("Don't know what " + part + " is");
```

File: api/src/main/java/net/kyori/adventure/title/TitlePart.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      return "TitlePart.TITLE";
+    }
+  };
+
+  /**
+   * The subtitle part of a title.
+   *
+   * @since 4.9.0
+   */
+  @NotNull TitlePart<Component> SUBTITLE = new TitlePart<Component>() {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/title/TitlePart.java:56]:
```suggestion
  TitlePart<Component> SUBTITLE = new TitlePart<Component>() {
```

File: api/src/main/java/net/kyori/adventure/title/TitlePart.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ * @since 4.9.0
+ */
+@ApiStatus.NonExtendable
+public interface TitlePart<T> {
+  /**
+   * A title part of a title.
+   *
+   * @since 4.9.0
+   */
+  @NotNull TitlePart<Component> TITLE = new TitlePart<Component>() {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/title/TitlePart.java:44]:
```suggestion
  TitlePart<Component> TITLE = new TitlePart<Component>() {
```

File: api/src/main/java/net/kyori/adventure/title/TitlePart.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      return "TitlePart.SUBTITLE";
+    }
+  };
+
+  /**
+   * The times part of a title.
+   *
+   * @since 4.9.0
+   */
+  @NotNull TitlePart<Title.Times> TIMES = new TitlePart<Title.Times>() {
```
Comment (kashike) [inline api/src/main/java/net/kyori/adventure/title/TitlePart.java:68]:
```suggestion
  TitlePart<Title.Times> TIMES = new TitlePart<Title.Times>() {
```

**Review by kashike — CHANGES_REQUESTED** [review]

**Review by kashike — APPROVED** [review]

**Pull request merged**